src/main/java/com/conveyal/r5/analyst/TravelTimeSurfaceComputer.java

```diff
@@ -0,0 +1,179 @@
+    private static final WebMercatorGridPointSetCache pointSetCache = new WebMercatorGridPointSetCache();
+
+    public static final int PERCENTILE_STEP = 5;
+
+    public final TravelTimeSurfaceRequest request;
+    public final TransportNetwork network;
+
+    public TravelTimeSurfaceComputer (TravelTimeSurfaceRequest request, TransportNetwork network) {
+        this.request = request;
+        this.network = network;
+    }
+
+    public void write (OutputStream os) throws IOException {
+        StreetMode accessMode = LegMode.legModeSetToDominantStreetMode(request.request.accessModes);
+        StreetMode directMode = LegMode.legModeSetToDominantStreetMode(request.request.directModes);
+
+        int nSamples = 100 / PERCENTILE_STEP - 1;
+
+        WebMercatorGridPointSet destinations = pointSetCache.get(request.zoom, request.west, request.north, request.width, request.height);
+
+        AccessGridWriter output;
+        try {
+            output = new AccessGridWriter(request.zoom, request.west, request.north, request.width, request.height, nSamples);
+        } catch (IOException e) {
+            // in memory, should not be able to throw this
+            throw new RuntimeException(e);
+        }
+
+        if (request.request.transitModes.isEmpty()) {
+            // non transit search
+            StreetRouter sr = new StreetRouter(network.streetLayer);
+            sr.timeLimitSeconds = request.request.maxTripDurationMinutes * 60;
+            sr.streetMode = directMode;
+            sr.dominanceVariable = StreetRouter.State.RoutingVariable.DURATION_SECONDS;
+            sr.profileRequest = request.request;
+            sr.setOrigin(request.request.fromLat, request.request.fromLon);
+            sr.route();
+
+            int offstreetTravelSpeedMillimetersPerSecond = (int) (request.request.getSpeed(directMode) * 1000);
+
+            LinkedPointSet linkedDestinations = destinations.link(network.streetLayer, directMode);
+
+            int[] travelTimesToTargets = linkedDestinations.eval(sr::getTravelTimeToVertex, offstreetTravelSpeedMillimetersPerSecond).travelTimes;
+            for (int target = 0; target < travelTimesToTargets.length; target++) {
+                int x = target % request.width;
+                int y = target / request.width;
+
+                final int travelTimeMinutes =
+                        travelTimesToTargets[target] == RaptorWorker.UNREACHED ? RaptorWorker.UNREACHED : travelTimesToTargets[target] / 60;
+                // the frontend expects percentiles of travel time. There is no variation in nontransit travel time so
+                // just replicate the same number repeatedly. This could be improved, but at least it will compress well.
+                // int divide (floor) used below as well. TODO is this wise?
+                int[] results = IntStream.range(0, nSamples).map(i -> travelTimeMinutes).toArray();
```

Comment:
Yes, but there is code in the frontend that depends on there being 20 percentiles, changing that is a pretty major project, I estimate it at at least a day. I can do it but probably not for at least a week and possibly not until I return from Australia.